node/bin/src/batcher/mod.rs

```diff
@@ -78,7 +79,56 @@ impl PipelineComponent for Batcher {
         let latency_tracker = ComponentStateReporter::global()
             .handle_for("batcher", GenericComponentState::WaitingRecv);
 
-        let mut prev_batch_info = self.startup_config.prev_batch_info.clone();
+        // First batch in the channel, if present, is always last executed batch. If it is missing
+        // (channel is empty), then there are no executed batches on L1 yet, and we should start from
+        // genesis.
+        let (mut prev_batch_info, first_expected_block) = match self.committed_batches.recv().await
+        {
+            None => {
+                tracing::info!("no committed batches yet; defaulting to genesis");
+                (self.startup_config.genesis_batch_info.clone(), 1)
+            }
+            Some(last_executed_batch) => {
+                tracing::info!(
+                    ?last_executed_batch,
+                    "using last executed batch as starting point"
+                );
+                let last_executed_block_number =
+                    last_executed_batch.commit_info.last_block_number.unwrap();
+                // todo: stop using this struct once fully migrated from S3
+                let last_executed_batch_info = BatchInfo {
+                    commit_info: last_executed_batch.commit_info,
+                    chain_address: Default::default(),
+                    upgrade_tx_hash: last_executed_batch.upgrade_tx_hash,
+                    blob_sidecar: None,
+                };
+
+                (
+                    last_executed_batch_info.into_stored(&last_executed_batch.protocol_version),
+                    last_executed_block_number + 1,
+                )
+            }
+        };
+
+        // We might receive some blocks that belong to already executed batches. We can skip these
+        // as there is no need to perform any L1 operations on them.
```

Comment:
hmm this makes it impossible to re-prove an already executed batch - but maybe indeed it's not such a important operations tool..